dotnet/private/rules/publish_binary/publish_binary.bzl

```diff
@@ -0,0 +1,376 @@
+    is_windows = ctx.target_platform_has_constraint(ctx.attr._windows_constraint[platform_common.ConstraintValueInfo])
+    inputs = [binary_info.app_host]
+    app_host_copy = ctx.actions.declare_file(
+        "{}/publish/{}/{}".format(ctx.label.name, runtime_identifier, binary_info.app_host.basename),
+    )
+    outputs = [app_host_copy]
+    script_body = []
+
+    _copy_file(script_body, binary_info.app_host, app_host_copy, is_windows = is_windows)
+
+    # All managed DLLs are copied next to the app host in the publish directory
+    for file in assembly_info.lib + assembly_info.transitive_lib.to_list():
+        output = ctx.actions.declare_file(
+            "{}/publish/{}/{}".format(ctx.label.name, runtime_identifier, file.basename),
+        )
+        outputs.append(output)
+        inputs.append(file)
+        _copy_file(script_body, file, output, is_windows = is_windows)
+
+    # When publishing a self-contained binary, we need to copy the native DLLs to the
+    # publish directory as well. If the binary is not self-contained, we need to copy
+    # the native files to a subfolder with the pattern `runtimes/<RID>/native/<file>`
+    # or `runtimes/<RID>/<TFM>/<file>`.
+    for file in assembly_info.native + assembly_info.transitive_native.to_list():
+        # TODO: Do correct folder structure for non-self-contained publishing
+        inputs.append(file)
+        output = ctx.actions.declare_file(
+            "{}/publish/{}/{}".format(ctx.label.name, runtime_identifier, file.basename),
+        )
+        outputs.append(output)
+        _copy_file(script_body, file, output, is_windows = is_windows)
+
+    # The data files put into the publish folder in a structure that works with
+    # the runfiles lib. End users should not expect files in the `data` attribute
+    # to be resolvable by relative paths. They need to use the runfiles lib.
+    #
+    # Since we want the published binary and all it's files to be easily extracted
+    # into e.g. a tar/zip/docker we manually create the runfiles structure because
+    # there are many sharp edges with extracting runfiles from Bazel. By manually
+    # creating the runfiles structure the runfiles are just normal files in the
+    # DefaultInfo provider and can thus be easily forwarded to filegroups/tars/containers.
+    #
+    # The runfiles library follows the spec and tries to find a `<DLL>.runfiles` directory
+    # next to the the DLL based on argv0 of the running process if
+    # RUNFILES_DIR/RUNFILES_MANIFEST_FILE/RUNFILES_MANIFEST_ONLY is not set).
+    for file in assembly_info.data + assembly_info.transitive_data.to_list():
+        inputs.append(file)
+        manifest_path = _to_manifest_path(ctx, file)
+        output = ctx.actions.declare_file(
+            "{}/publish/{}/{}.runfiles/{}".format(ctx.label.name, runtime_identifier, binary_info.app_host.basename, manifest_path),
```

Comment:
This looks quite scary. Are you reconstructing the runfiles folder. Doesnt this collide with the native bazel runfiles symlink tree? I think we should try to match msbuild publish behavior and just copy all binaries to the root publish folder except for runtime specific ones.

Reply:
I just ran the example but don't see the .runfiles folder in the output. Can you explain why it's mentioned here?

Reply:
O I see the pdb file is written there.

Reply:
The reason or this is that runfiles have generally had a bad track record with packaging rules
See https://github.com/bazelbuild/rules_pkg/pull/605 for some discussion. 

The only files that are put into the runfiles folder are the items that go into `data` on each rule. This is so that the runfiles library can find the data files under the expected location.
All other files follow the standard .Net convention where everything is copied to the root of the publish folder.

I agree that this isn't very orthodox but this makes it possible to publish  the applications without having to wrangle possible runfile issues when you e.g. want to zip/tar/containerize the output of this rule. I can try using the standard runfiles once the runfiles support lands in rules_pkg.

Reply:
Regarding the PDB files, I was thinking about adding a specific attribute for them in the providers and not add them to `data`. That way they would just also end up in the root publishing folder.

Reply:
Hmm I dont think I like this approach. We want to be as close as possible to msbuild logic. So when we have some assets flagged as copy to output, they end up in data[]. They should be copied to the publish folder without the runfiles subdir since this is not a msbuild concept. I'm not sure how msbuild handles directory structures here, but we should try to copy that behavior.

Reply:
The `data` attribute is very Bazel specific so I don't think we should conflate it with the `copy to output` functionality of MSBuild. If we want to add that functionality we should add another `copy_to_output` attribute that e.g. takes in a map of output pathts and file labels.

We don't want to break the Bazel conventions as well which is very well defined for the `data` attribute.

Reply:
hmm you mean the magic propagation of data attribute values? I'm not quite sure about it's use case trough bazel conventions but I have read about it. We currently use data[] to propagate the copy to output files, for example for assets that used during tests.

Reply:
A separate copy_to_output sounds reasonable

Reply:
Yeah, I agree that the `data` attribute can be quite obscure to those who are not used to Bazel but I don't think we should break the set convention where the `data` files should be part of the runfiles.

We could also add an attribute `copy_data_to_outputs` which copies the data files to the output folder while keeping the relative folder structure.